micro_http/src/server.rs

```diff
@@ -0,0 +1,681 @@
+/// server runs on incapable of performing other operations, or it can
+/// be used in another `EPOLL` structure, as it provides its `epoll_fd`,
+/// the file descriptor of the epoll structure used within the server,
+/// and it can be added to another one using the `EPOLLIN` flag. Whenever
+/// there is a notification on that fd, `handle_notifications` should be
+/// called once.
+pub struct HttpServer {
+    /// Socket on which we listen for new connections.
+    socket: UnixListener,
+    /// File descriptor of the server's epoll structure.
+    epoll_fd: RawFd,
+    /// Holds the token-connection pairs of the server.
+    /// Each connection has an associated identification token.
+    connections: HashMap<RawFd, ClientConnection<UnixStream>>,
+}
+
+impl HttpServer {
+    /// Constructor for `HttpServer`.
+    pub fn new<P: AsRef<Path>>(path_to_socket: P) -> Result<Self> {
+        let socket = UnixListener::bind(path_to_socket).map_err(ServerError::IOError)?;
+        let epoll_fd = epoll::create(true).map_err(ServerError::IOError)?;
+        Ok(HttpServer {
+            socket,
+            epoll_fd,
+            connections: HashMap::new(),
+        })
+    }
+
+    /// Starts the HTTP Server.
+    pub fn start_server(&mut self) -> Result<()> {
+        // Add the socket on which we listen for new connections to the
+        // `epoll` structure.
+        Self::epoll_add(self.epoll_fd, self.socket.as_raw_fd())
+    }
+
+    /// Calls `epoll::wait` on the epoll structure containing all open connections.
+    ///
+    /// This function is responsible for the data exchange with the clients and should
+    /// be called when we are either notified through `epoll` that we need to exchange
+    /// data with at least a client or when we don't need to perform any other operations
+    /// on this thread and we can afford to call it in a loop.
+    ///
+    /// Returns a collection of complete and valid requests to be processed by the user
+    /// of the server. Once processed, responses should be sent using `enqueue_responses()`.
+    pub fn handle_notifications(&mut self) -> Result<Vec<RequestWithID>> {
+        let mut parsed_requests: Vec<RequestWithID> = vec![];
+        let mut events = vec![epoll::Event::new(epoll::Events::empty(), 0); MAX_CONNECTIONS];
+        let epoll_fd = self.epoll_fd;
+        let num_ev = epoll::wait(epoll_fd, -1, &mut events[..]).map_err(ServerError::IOError)?;
+        for e in events.iter().take(num_ev) {
+            // We have received a new connection on the server socket.
+            if e.data as RawFd == self.socket.as_raw_fd() {
+                match self.handle_new_connection() {
+                    // If the server is full, we send a message to the client
+                    // notifying him that we will close the connection, then
+                    // we discard it.
+                    Err(ServerError::ServerFull) => continue,
```

Comment:
I prefer to handle the error case you mention in the comments in this function and not in `handle_notifications`.